airbyte-integrations/connectors/source-snowflake/src/main/java/io.airbyte.integrations.source.snowflake/SnowflakeDataSourceUtils.java

```diff
@@ -0,0 +1,168 @@
+import org.slf4j.LoggerFactory;
+
+public class SnowflakeDataSourceUtils {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SnowflakeDataSourceUtils.class);
+  private static final int PAUSE_BETWEEN_TOKEN_REFRESH_MIN = 7; // snowflake access token's TTL is 10min and can't be modified
+  private static final String REFRESH_TOKEN_URL = "https://%s/oauth/token-request";
+  private static final HttpClient httpClient = HttpClient.newBuilder()
+      .version(HttpClient.Version.HTTP_2)
+      .connectTimeout(Duration.ofSeconds(10))
+      .build();
+
+  /**
+   * Snowflake OAuth access token expires in 10 minutes. For the cases when sync duration is more than
+   * 10 min, it requires updating 'token' property after the start of connection pool.
+   * HikariDataSource brings support for this requirement.
+   *
+   * @param config source config JSON
+   * @return datasource
+   */
+  public static HikariDataSource createDataSource(final JsonNode config) {
+    HikariConfig hikariConfig = new HikariConfig();
+    final StringBuilder jdbcUrl = new StringBuilder(
+        String.format("jdbc:snowflake://%s/?", config.get("host").asText()));
+    jdbcUrl.append(String.format(
+        "role=%s&warehouse=%s&database=%s&schema=%s&JDBC_QUERY_RESULT_FORMAT=%s&CLIENT_SESSION_KEEP_ALIVE=%s",
+        config.get("role").asText(),
+        config.get("warehouse").asText(),
+        config.get("database").asText(),
+        config.get("schema").asText(),
+        // Needed for JDK17 - see
+        // https://stackoverflow.com/questions/67409650/snowflake-jdbc-driver-internal-error-fail-to-retrieve-row-count-for-first-arrow
+        "JSON",
+        true));
+    if (config.has("jdbc_url_params")) {
+      jdbcUrl.append(config.get("jdbc_url_params").asText());
+    }
+    hikariConfig.setJdbcUrl(jdbcUrl.toString());
+
+    if (config.has("credentials") && config.get("credentials").has("auth_type")
+        && "OAuth".equals(config.get("credentials").get("auth_type").asText())) {
+      LOGGER.info("Authorization mode is OAuth");
+      try {
+        var credentials = config.get("credentials");
+        Properties properties = new Properties();
+        properties.setProperty("client_id", credentials.get("client_id").asText());
+        properties.setProperty("client_secret", credentials.get("client_secret").asText());
+        properties.setProperty("refresh_token", credentials.get("refresh_token").asText());
+        properties.setProperty("host", config.get("host").asText());
+        var accessToken = getAccessTokenUsingRefreshToken(
+            config.get("host").asText(), credentials.get("client_id").asText(),
+            credentials.get("client_secret").asText(), credentials.get("refresh_token").asText());
+        properties.put("authenticator", "oauth");
+        properties.put("token", accessToken);
+        properties.put("account", config.get("host").asText());
+        hikariConfig.setDataSourceProperties(properties);
```

Comment:
can we move properties populating block into a separate method to make it more readable?

Reply:
done